core/state/state_object.go

```diff
@@ -436,6 +452,13 @@ func (s *StateObject) updateTrie(db Database) Trie {
 
 // UpdateRoot sets the trie root to the current root hash of
 func (s *StateObject) updateRoot(db Database) {
+	// If node runs in no trie mode, set root to empty.
+	defer func() {
+		if db.NoTries() {
+			s.data.Root = common.Hash{}
+		}
+	}()
```

Comment:
should we call `updateTrie` below if trie is not enabled

Reply:
This is a deferred function that normally executes `updateTrie` below. If `NoTries` is enabled, root is set to `empty`

Reply:
I guess you try to keep the behavior unchanged, by using the defer(){...}, right?
It is ok to me, since the current logic of `updateRoot` is a bit completed.